File: backend/src/services/ozowService.js
const crypto = require('crypto');
const axios = require('axios');
const admin = require('../../firebase');

/**
 * Ozow Pay-by-Bank Integration Service
 * Based on Ozow Hub documentation: https://hub.ozow.com/
 */
class OzowService {
  constructor() {
    this.config = {
      siteCode: String(process.env.OZOW_SITE_CODE || '').trim(),
      privateKey: String(process.env.OZOW_PRIVATE_KEY || '').trim(),
      apiKey: String(process.env.OZOW_API_KEY || '').trim(),
      isTest: process.env.OZOW_IS_TEST === 'true',
      successUrl: String(process.env.OZOW_SUCCESS_URL || '').trim(),
      cancelUrl: String(process.env.OZOW_CANCEL_URL || '').trim(),
      errorUrl: String(process.env.OZOW_ERROR_URL || '').trim(),
      notifyUrl: String(process.env.OZOW_NOTIFY_URL || '').trim(),
    };

    // Ozow API endpoints
    this.paymentUrl = this.config.isTest
      ? 'https://stagingpay.ozow.com'
      : 'https://pay.ozow.com';

    this.apiUrl = this.config.isTest
      ? 'https://stagingapi.ozow.com'
      : 'https://api.ozow.com';
  }

  /**
   * Generate SHA512 hash for Ozow request verification.
   * Ozow requires: concatenate specific fields in order + PrivateKey, then SHA512, lowercase.
   * 
   * Fields in order for payment request hash:
   * SiteCode + CountryCode + CurrencyCode + Amount + TransactionReference + 
   * BankReference + Optional3 + CancelUrl + ErrorUrl + SuccessUrl + NotifyUrl + IsTest + PrivateKey
   */
  generateRequestHash(data) {
    const hashString = [
      data.SiteCode,
      data.CountryCode,
      data.CurrencyCode,
      data.Amount,
      data.TransactionReference,
      data.BankReference,
      data.Optional1 || '',
      data.Optional2 || '',
      data.Optional3 || '',
      data.Optional4 || '',
      data.Optional5 || '',
      data.CancelUrl,
      data.ErrorUrl,
      data.SuccessUrl,
      data.NotifyUrl,
      data.IsTest,
      this.config.privateKey,
    ].join('');

    const hash = crypto.createHash('sha512').update(hashString.toLowerCase()).digest('hex');

    console.log('\n=== Ozow Hash Generation ===');
    console.log('Hash input (redacted key):', hashString.replace(this.config.privateKey, '***'));
    console.log('Generated hash:', hash);
    console.log('============================\n');

    return hash;
  }

  /**
   * Generate SHA512 hash for notification verification.
   * 
   * Fields for notification hash:
   * SiteCode + TransactionId + TransactionReference + Amount + Status + 
   * Optional1 + Optional2 + Optional3 + Optional4 + Optional5 + CurrencyCode + 
   * IsTest + StatusMessage + PrivateKey
   */
  generateNotificationHash(data) {
    const hashString = [
      data.SiteCode,
      data.TransactionId,
      data.TransactionReference,
      data.Amount,
      data.Status,
      data.Optional1 || '',
      data.Optional2 || '',
      data.Optional3 || '',
      data.Optional4 || '',
      data.Optional5 || '',
      data.CurrencyCode,
      data.IsTest,
      data.StatusMessage || '',
      this.config.privateKey,
    ].join('');

    return crypto.createHash('sha512').update(hashString.toLowerCase()).digest('hex');
  }

  /**
   * Creates payment request data for Ozow.
   * Returns URL + form fields to POST to Ozow.
   */
  createPaymentRequest(orderData, orderId, userId) {
    try {
      if (!this.config.siteCode || !this.config.privateKey) {
        throw new Error('Ozow SiteCode or PrivateKey is not configured.');
      }

      const amountString = parseFloat(orderData.total || 0).toFixed(2);

      if (parseFloat(amountString) <= 0) {
        throw new Error('Invalid order amount. Must be greater than zero.');
      }

      // Build Ozow payment request data
      const paymentData = {
        SiteCode: this.config.siteCode,
        CountryCode: 'ZA',
        CurrencyCode: 'ZAR',
        Amount: amountString,
        TransactionReference: orderId,
        BankReference: `LZA-${orderId.slice(-8)}`,
        Optional1: userId,
        Optional2: orderData.email || '',
        Optional3: '',
        Optional4: '',
        Optional5: '',
        CancelUrl: `${this.config.cancelUrl}/${orderId}`,
        ErrorUrl: `${this.config.errorUrl}/${orderId}`,
        SuccessUrl: `${this.config.successUrl}/${orderId}`,
        NotifyUrl: this.config.notifyUrl,
        IsTest: this.config.isTest.toString().toLowerCase(),
      };

      // Generate hash
      const hashCheck = this.generateRequestHash(paymentData);
      paymentData.HashCheck = hashCheck;

      console.log('=== Ozow Payment Request Summary ===');
      console.log('Payment data:', { ...paymentData, HashCheck: hashCheck.substring(0, 20) + '...' });
      console.log('=====================================');

      return {
        formData: paymentData,
        url: this.paymentUrl,
        paymentId: orderId,
      };
    } catch (error) {
      console.error('Error creating Ozow payment request:', error);
      throw error;
    }
  }

  /**
   * Verify notification hash from Ozow callback.
   */
  verifyNotificationHash(data) {
    try {
      const calculatedHash = this.generateNotificationHash(data);
      const receivedHash = (data.Hash || data.HashCheck || '').toLowerCase();

      console.log('=== Ozow Notification Hash Verification ===');
      console.log('Received:', receivedHash);
      console.log('Calculated:', calculatedHash);
      console.log('Match:', calculatedHash === receivedHash);
      console.log('============================================');

      return calculatedHash === receivedHash;
    } catch (error) {
      console.error('Hash verification error:', error);
      return false;
    }
  }

  /**
   * Check transaction status via Ozow API (Step 3 - recommended).
   * GET /GetTransactionByReference?siteCode={siteCode}&transactionReference={ref}
   */
  async checkTransactionStatus(transactionReference) {
    try {
      const response = await axios.get(
        `${this.apiUrl}/GetTransactionByReference`,
        {
          params: {
            siteCode: this.config.siteCode,
            transactionReference: transactionReference,
          },
          headers: {
            Accept: 'application/json',
            ApiKey: this.config.apiKey,
          },
          timeout: 10000,
        }
      );

      console.log('Ozow status check response:', response.data);
      return response.data;
    } catch (error) {
      console.error('Ozow status check error:', error.message);
      return null;
    }
  }

  /**
   * Process Ozow notification callback.
   * 
   * Ozow notification statuses:
   * - "Complete" = payment successful
   * - "Cancelled" = user cancelled
   * - "Error" = payment error
   * - "Abandoned" = user abandoned
   * - "PendingInvestigation" = under review
   */
  async processNotification(data) {
    try {
      console.log('=== Processing Ozow Notification ===');
      console.log('TransactionReference:', data.TransactionReference);
      console.log('Status:', data.Status);
      console.log('TransactionId:', data.TransactionId);

      // Step 1: Verify hash
      if (!this.verifyNotificationHash(data)) {
        console.error('Invalid hash');
        return { success: false, error: 'Invalid hash' };
      }

      // Step 2: Verify site code matches
      if (data.SiteCode !== this.config.siteCode) {
        console.error('Site code mismatch');
        return { success: false, error: 'Site code mismatch' };
      }

      // Step 3 (recommended): Verify with Ozow API
      const apiStatus = await this.checkTransactionStatus(data.TransactionReference);
      if (apiStatus && apiStatus.status && 
          apiStatus.status.toLowerCase() !== data.Status.toLowerCase()) {
        console.error('Status mismatch between notification and API');
        return { success: false, error: 'Status mismatch' };
      }

      const orderId = data.TransactionReference;
      if (!orderId) {
        return { success: false, error: 'Missing order ID (TransactionReference)' };
      }

      // Update order in database
      const orderRef = admin.firestore().collection('orders').doc(orderId);
      const orderDoc = await orderRef.get();

      if (!orderDoc.exists) {
        return { success: false, error: 'Order not found' };
      }

      // Log notification for audit
      await admin.firestore().collection('payment_notifications').add({
        orderId,
        paymentData: data,
        receivedAt: admin.firestore.FieldValue.serverTimestamp(),
        provider: 'ozow',
        environment: this.config.isTest ? 'staging' : 'production',
        verified: true,
      });

      // Map Ozow status to your order status
      const ozowStatus = (data.Status || '').toLowerCase();
      const updateData = {
        paymentData: data,
        paymentProvider: 'ozow',
        paymentUpdatedAt: admin.firestore.FieldValue.serverTimestamp(),
        paymentVerified: true,
      };

      if (ozowStatus === 'complete') {
        updateData.status = 'pending';
        updateData.paymentStatus = 'paid';
        updateData.paymentCompleted = true;
        updateData.paymentCompletedAt = admin.firestore.FieldValue.serverTimestamp();
        updateData.ozow_transaction_id = data.TransactionId;
        updateData.transaction_id = data.TransactionId;
      } else if (ozowStatus === 'cancelled' || ozowStatus === 'abandoned') {
        updateData.status = 'cancelled';
        updateData.paymentStatus = 'cancelled';
      } else if (ozowStatus === 'error') {
        updateData.status = 'payment_failed';
        updateData.paymentStatus = 'failed';
      } else if (ozowStatus === 'pendinginvestigation') {
        updateData.status = 'pending_payment';
        updateData.paymentStatus = 'pending_investigation';
      }

      await orderRef.update(updateData);

      console.log('=== Notification processed successfully ===');

      return {
        success: true,
        orderId,
        status: data.Status,
        newOrderStatus: updateData.status,
        userId: orderDoc.data().userId,
      };
    } catch (error) {
      console.error('Ozow notification processing error:', error);
      return { success: false, error: error.message };
    }
  }
}

module.exports = new OzowService();

File: api/Services/OzowService.cs
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using LocalsZaApi.Models;

namespace LocalsZaApi.Services;

public class OzowService
{
    private readonly IConfiguration _config;
    private readonly ILogger<OzowService> _log;
    private readonly IHttpClientFactory _httpFactory;

    public OzowService(IConfiguration config, ILogger<OzowService> log, IHttpClientFactory httpFactory)
    {
        _config = config;
        _log = log;
        _httpFactory = httpFactory;
    }

    private string SiteCode   => _config["Ozow:SiteCode"]   ?? throw new InvalidOperationException("Ozow:SiteCode missing");
    private string PrivateKey => _config["Ozow:PrivateKey"] ?? throw new InvalidOperationException("Ozow:PrivateKey missing");
    private string ApiKey     => _config["Ozow:ApiKey"]     ?? throw new InvalidOperationException("Ozow:ApiKey missing");
    private bool   IsTest     => bool.Parse(_config["Ozow:IsTest"] ?? "true");
    private string PaymentUrl => _config["Ozow:PaymentUrl"] ?? "https://stagingpay.ozow.com";
    private string ApiUrl     => _config["Ozow:ApiUrl"]     ?? "https://stagingapi.ozow.com";

    /// <summary>
    /// Build the payload + URL the frontend should POST to (Ozow consumer checkout).
    /// </summary>
    public OzowInitiateResult Initiate(Order order, string? customerEmail, string baseAppUrl)
    {
        var amount = order.Total.ToString("F2", CultureInfo.InvariantCulture);

        if (decimal.Parse(amount, CultureInfo.InvariantCulture) <= 0)
            throw new InvalidOperationException("Order amount must be greater than zero");

        // BankReference appears on the customer's bank statement — keep readable, max 20 chars.
        // Mirrors JS: LZA-<last 8 of orderId>
        var tail = order.Id.Length <= 8 ? order.Id : order.Id[^8..];
        var bankRef = $"LZA-{tail}";

        var payload = new Dictionary<string, string>
        {
            ["SiteCode"]             = SiteCode,
            ["CountryCode"]          = "ZA",
            ["CurrencyCode"]         = "ZAR",
            ["Amount"]               = amount,
            ["TransactionReference"] = order.Id,
            ["BankReference"]        = bankRef,
            ["Optional1"]            = order.UserId ?? order.GuestId ?? "",
            ["Optional2"]            = customerEmail ?? "",
            ["Optional3"]            = "",
            ["Optional4"]            = "",
            ["Optional5"]            = "",
            ["CancelUrl"]            = $"{baseAppUrl}/order/payment/cancelled/{order.Id}",
            ["ErrorUrl"]             = $"{baseAppUrl}/order/payment/error/{order.Id}",
            ["SuccessUrl"]           = $"{baseAppUrl}/order/payment/success/{order.Id}",
            ["NotifyUrl"]            = $"{baseAppUrl}/api/payment/notify",
            ["IsTest"]               = IsTest.ToString().ToLowerInvariant(),
        };

        var hash = GenerateRequestHash(payload);
        payload["HashCheck"] = hash;

        _log.LogInformation("Ozow initiate: orderId={OrderId} amount={Amount} bankRef={BankRef}",
            order.Id, amount, bankRef);

        return new OzowInitiateResult(PostUrl: PaymentUrl, Fields: payload, Hash: hash);
    }

    /// <summary>
    /// Verify a notification from Ozow's webhook by recomputing the hash.
    /// </summary>
    public bool VerifyNotification(IReadOnlyDictionary<string, string> form)
    {
        // Ozow may send "Hash" or (rarely) "HashCheck"
        var received = (form.GetValueOrDefault("Hash")
                      ?? form.GetValueOrDefault("HashCheck")
                      ?? "").ToLowerInvariant();

        var computed = GenerateNotificationHash(form);
        var match = string.Equals(computed, received, StringComparison.OrdinalIgnoreCase);

        if (!match)
            _log.LogWarning("Ozow notify hash mismatch. received={Received} computed={Computed}",
                received, computed);

        return match;
    }

    /// <summary>
    /// Step 3 (recommended) — query Ozow's API to confirm the notification's status.
    /// Protects against spoofed webhook requests even if hash somehow passes.
    /// </summary>
    public async Task<OzowApiStatus?> CheckTransactionStatusAsync(string transactionReference)
    {
        try
        {
            var http = _httpFactory.CreateClient();
            http.Timeout = TimeSpan.FromSeconds(10);

            var url = $"{ApiUrl}/GetTransactionByReference"
                    + $"?siteCode={Uri.EscapeDataString(SiteCode)}"
                    + $"&transactionReference={Uri.EscapeDataString(transactionReference)}";

            var req = new HttpRequestMessage(HttpMethod.Get, url);
            req.Headers.Add("Accept", "application/json");
            req.Headers.Add("ApiKey", ApiKey);

            var response = await http.SendAsync(req);
            if (!response.IsSuccessStatusCode)
            {
                _log.LogWarning("Ozow status API returned {Code}", response.StatusCode);
                return null;
            }

            var json = await response.Content.ReadAsStringAsync();
            // The endpoint returns an array — newest first
            var result = System.Text.Json.JsonSerializer.Deserialize<OzowApiStatus[]>(json,
                new System.Text.Json.JsonSerializerOptions { PropertyNameCaseInsensitive = true });

            return result?.FirstOrDefault();
        }
        catch (Exception ex)
        {
            _log.LogError(ex, "Ozow status check failed");
            return null;
        }
    }

    // ---------- hash helpers ----------

    /// <summary>
    /// Request hash field order (matches our working JS impl):
    /// SiteCode + CountryCode + CurrencyCode + Amount + TransactionReference + BankReference +
    /// Optional1..Optional5 + CancelUrl + ErrorUrl + SuccessUrl + NotifyUrl + IsTest + PrivateKey
    /// </summary>
    private string GenerateRequestHash(Dictionary<string, string> p)
    {
        var sb = new StringBuilder();
        sb.Append(p["SiteCode"]);
        sb.Append(p["CountryCode"]);
        sb.Append(p["CurrencyCode"]);
        sb.Append(p["Amount"]);
        sb.Append(p["TransactionReference"]);
        sb.Append(p["BankReference"]);
        sb.Append(p["Optional1"]);
        sb.Append(p["Optional2"]);
        sb.Append(p["Optional3"]);
        sb.Append(p["Optional4"]);
        sb.Append(p["Optional5"]);
        sb.Append(p["CancelUrl"]);
        sb.Append(p["ErrorUrl"]);
        sb.Append(p["SuccessUrl"]);
        sb.Append(p["NotifyUrl"]);
        sb.Append(p["IsTest"]);
        sb.Append(PrivateKey);

        return Sha512(sb.ToString().ToLowerInvariant());
    }

    /// <summary>
    /// Notification hash field order (matches our working JS impl):
    /// SiteCode + TransactionId + TransactionReference + Amount + Status +
    /// Optional1..Optional5 + CurrencyCode + IsTest + StatusMessage + PrivateKey
    /// </summary>
    private string GenerateNotificationHash(IReadOnlyDictionary<string, string> n)
    {
        string V(string k) => n.GetValueOrDefault(k) ?? "";

        var sb = new StringBuilder();
        sb.Append(V("SiteCode"));
        sb.Append(V("TransactionId"));
        sb.Append(V("TransactionReference"));
        sb.Append(V("Amount"));
        sb.Append(V("Status"));
        sb.Append(V("Optional1"));
        sb.Append(V("Optional2"));
        sb.Append(V("Optional3"));
        sb.Append(V("Optional4"));
        sb.Append(V("Optional5"));
        sb.Append(V("CurrencyCode"));
        sb.Append(V("IsTest"));
        sb.Append(V("StatusMessage"));
        sb.Append(PrivateKey);

        return Sha512(sb.ToString().ToLowerInvariant());
    }

    private static string Sha512(string input)
    {
        var bytes = Encoding.UTF8.GetBytes(input);
        var hash  = SHA512.HashData(bytes);
        return Convert.ToHexString(hash).ToLowerInvariant();
    }
}

public record OzowInitiateResult(string PostUrl, IDictionary<string, string> Fields, string Hash);

public class OzowApiStatus
{
    public string? TransactionId        { get; set; }
    public string? TransactionReference { get; set; }
    public string? Status               { get; set; }
    public string? StatusMessage        { get; set; }
    public decimal? Amount              { get; set; }
}